src/flow_field.py

```diff
@@ -12,20 +12,44 @@
 
 # See https://floris.readthedocs.io for documentation
 
+from __future__ import annotations
 
+import attr
 import numpy as np
+import numpy.typing as npt
+
+from src.utilities import FromDictMixin, int_attrib, float_attrib, attrs_array_converter
 
 from .grid import Grid
 
 
-class FlowField:
-    def __init__(self, input_dictionary):
-        self.wind_shear = input_dictionary["wind_shear"]
-        self.wind_veer = input_dictionary["wind_veer"]
-        self.wind_speeds = np.array(input_dictionary["wind_speeds"])
-        self.wind_directions = np.array(input_dictionary["wind_directions"])
-        self.reference_wind_height = input_dictionary["reference_wind_height"]
-        self.air_density = input_dictionary["air_density"]
+NDArrayFloat = npt.NDArray[np.float64]
+
+
+@attr.s(auto_attribs=True)
+class FlowField(FromDictMixin):
+    wind_shear: float = float_attrib()
+    wind_veer: float = float_attrib()
+    wind_speeds: NDArrayFloat = attr.ib(converter=attrs_array_converter)
+    wind_directions: NDArrayFloat = attr.ib(converter=attrs_array_converter)
+    reference_wind_height: int = int_attrib()
+    reference_turbine_diameter: float = float_attrib()
+    air_density: float = float_attrib()
+
+    n_wind_speeds: int = attr.ib(init=False)
```

Comment:
What does it mean if these are defined as attributes here and included as virtual properties below?

Reply:
I think that's a mixing with the two PRs because I had originally removed the virtual property. A solution could be to have these properties update if the coordinates are updated, but in general, I'm skeptical of updating the fields after initialization.

Reply:
See my comment below for more on this.